### DISCUSSION TITLE
Top Level Await on the client

### DISCUSSION THREAD

**Discussion by matheusccastroo** [meteor/meteor#12109] [Feature Requests]
Hello! How's everyone doing? Hope all is well.

We need to discuss an important point for the top level await feature: the legacy bundle. What we need to define is how to approach top level await for the legacy bundle, since such browsers may or may not support `promises/async functions` (for example, IE).
Right now on the [PR](https://github.com/meteor/meteor/pull/12095) we don't add this support even for what we call as `modernBrowsers` on the code, just the server currently supports top level await.

To proceed here, I think we have three options:

1. Drop support for legacy browsers: this is a bit harsh, but it's a good option too. We could just drop support for legacy browsers that don't support `promises/async functions`. You can check the affected browsers [here](https://caniuse.com/?search=async%20await), but basically would be IE, Opera Mini and anything else before 2016.
2. Run Babel again on the files/modules that uses Top Level Await when bundling for the legacy bundle: this brings some issues related to the build performance, but if we want to actually support this, it would be the way to go. Basically we would scope such additional pass through Babel just for the promises plugin and just for the legacy bundle. I also think here we could remove the legacy bundle from the default build for `meteor run` and `meteor build`.
3. Make the linker add the `promise/async` related code in the "transpiled" way: @zodern suggested that we could also output the code that will work on all browsers directly on the linker, removing the need to pass Babel again.

So yeah, this is our current options for now, and we would like to hear your thoughts as well!

If you have any opinion/suggestions, please let me know.

**Comment by zodern** [comment]
> @zodern suggested that we could also output the code that will work on all browsers directly on the linker, removing the need to pass Babel again.

I don't think this will be that difficult. Async/await is a nicer syntax for using promises, but we can do everything with promises directly. The `Promise` global won't always be available, but if the package or app uses TLA then it will be, though I don't think we will need to use it.

**Reply by fredmaiaarantes** [reply]
Option 3 looks like a great solution to me. We wouldn't drop the support for these browsers and wouldn't increase the build time, right?

**Reply by matheusccastroo** [reply]
@zodern But the issue still continues, right? If the browser doesn't support `async/await` then it probably won't support promises too. In this case we would still need to "polyfill" it, no?

While this option would solve both issues, it would make the code around this part more complex than necessary, which IMHO is not a worth trade-off.

**Reply by zodern** [reply]
> But the issue still continues, right? If the browser doesn't support async/await then it probably won't support promises too. In this case we would still need to "polyfill" it, no?

To explain what I meant, I'm going to group code into two types:
1. App or package code
2. Linker code (code generated by the linker that wasn't directly provided by an app or package).

Apps and packages can use promises, but if they want to work on legacy browsers they need to ensure a promise polyfill is used. Usually you do this by adding `ecmascript` or a similar package as a dependency. Top level await will be implemented as part of the ecmascript and typescript packages, so promises will always be available at the same time.

For linker code, Meteor doesn't really know at build time if promises will be available when the generated code runs on the legacy client.  Some packages will load before any promise polyfill (though these should only be ones that don't use promises). You can check if an object is a promise and use the promise without having a promise polyfill, so I don't think this will create any real difficulties for us. 

> While this option would solve both issues, it would make the code around this part more complex than necessary, which IMHO is not a worth trade-off.

After integrating the top level await implementation I've been working on in reify, and making the necessary changes to support the client, the only await that should be left is [this one](https://github.com/meteor/meteor/pull/12095/files#diff-0972e87d5bf49b0ff289a29dee4ae83bb04b83d2e057976f7a7680e0e1ee2e0aR471-R475). We can add a small helper to the `Package` object to handle the eager requires and any promises they return.
